orca-clouddriver/src/test/java/com/netflix/spinnaker/orca/clouddriver/pipeline/providers/gce/GceDeployStagePreProcessorTest.java

```diff
@@ -0,0 +1,132 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.when;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.netflix.spinnaker.orca.clouddriver.pipeline.providers.aws.ApplySourceServerGroupCapacityStage;
+import com.netflix.spinnaker.orca.clouddriver.pipeline.providers.aws.CaptureSourceServerGroupCapacityTask;
+import com.netflix.spinnaker.orca.clouddriver.pipeline.servergroup.ResizeServerGroupStage;
+import com.netflix.spinnaker.orca.clouddriver.pipeline.servergroup.strategies.DeployStagePreProcessor.StageDefinition;
+import com.netflix.spinnaker.orca.clouddriver.pipeline.servergroup.strategies.DeployStagePreProcessor.StepDefinition;
+import com.netflix.spinnaker.orca.clouddriver.pipeline.servergroup.support.TargetServerGroup;
+import com.netflix.spinnaker.orca.clouddriver.pipeline.servergroup.support.TargetServerGroupResolver;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class GceDeployStagePreProcessorTest {
+  private GceDeployStagePreProcessor preProcessor;
+  private ApplySourceServerGroupCapacityStage applySourceServerGroupCapacityStage;
+  private ResizeServerGroupStage resizeServerGroupStage;
+
+  @Mock private TargetServerGroupResolver targetServerGroupResolver;
+
+  @BeforeEach
+  void setUp() {
+    applySourceServerGroupCapacityStage = new ApplySourceServerGroupCapacityStage();
+    resizeServerGroupStage = new ResizeServerGroupStage();
+    preProcessor =
+        new GceDeployStagePreProcessor(
+            applySourceServerGroupCapacityStage, resizeServerGroupStage, targetServerGroupResolver);
+  }
+
+  @Test
+  void redBlackStrategyTest() {
+    when(targetServerGroupResolver.resolve(any()))
+        .thenReturn(
+            ImmutableList.of(
+                new TargetServerGroup(
+                    ImmutableMap.of("name", "testapp-v000", "zone", "us-central1-f"))));
+    Stage stage = new Stage();
+    Map<String, Object> context = new HashMap<>();
+    context.put("strategy", "redblack");
+    context.put("cloudProvider", "gce");
+    context.put("zone", "us-central1-f");
+    context.put(
+        "availabilityZones",
+        Collections.singletonMap("us-central1-f", Collections.singletonList("us-central1-f")));
+    context.put("application", "testapp");
+    stage.setContext(context);
+
+    // Before Stages
+    List<StageDefinition> beforeStages = preProcessor.beforeStageDefinitions(stage);
+    assertThat(beforeStages.stream().map(stageDefinition -> stageDefinition.stageDefinitionBuilder))
```

Comment:
I actually found a better (IMO) way to do this, but I didn't go back and fix the older tests I wrote:

```
assertThat(beforeStages)
    .extracting(stage -> stage.stageDefinitionBuilder)
    .containsExactly(resizeServerGroupStage)
```

Reply:
Wow this is great, I now love `extracting`! :sweet_potato: